pkg/tsdb/postgres/postgres.go

```diff
@@ -61,15 +86,161 @@ func escape(input string) string {
 }
 
-func generateConnectionString(datasource *models.DataSource, logger log.Logger) (string, error) {
+type certFileType int
+
+const (
+	rootCert = iota
+	clientCert
+	clientKey
+)
+
+func (t certFileType) String() string {
+	switch t {
+	case rootCert:
+		return "root certificate"
+	case clientCert:
+		return "client certificate"
+	case clientKey:
+		return "client key"
+	default:
+		panic(fmt.Sprintf("Unrecognized certFileType %d", t))
+	}
+}
+
+// writeCertFile writes a certificate file.
+func (s *postgresService) writeCertFile(ds *models.DataSource, fileContent, dataDir string, fileType certFileType) error {
+	var jsonFieldName string
+	var filename string
+	switch fileType {
+	case rootCert:
+		jsonFieldName = "generatedTLSRootCertFile"
+		filename = "root.crt"
+	case clientCert:
+		jsonFieldName = "generatedTLSCertFile"
+		filename = "client.crt"
+	case clientKey:
+		jsonFieldName = "generatedTLSKeyFile"
+		filename = "client.key"
+	default:
+		panic(fmt.Sprintf("unrecognized certFileType %s", fileType.String()))
+	}
+
+	s.logger.Debug("Writing cert file", "type", fileType, "jsonFieldName", jsonFieldName)
+
+	var generatedFilePath string
+	if tlsjs, ok := ds.JsonData.CheckGet(jsonFieldName); ok {
+		// Get generated file path from data source
+		generatedFilePath = tlsjs.MustString("")
+		s.logger.Debug("Got file path from data source", "path", generatedFilePath)
+	} else {
+		s.logger.Debug("File path not stored in data source")
+	}
+	if fileContent != "" {
+		if generatedFilePath == "" {
+			// Use standard filename
+			generatedFilePath = filepath.Join(dataDir, filename)
+		}
+
+		s.logger.Debug("Writing cert file", "path", generatedFilePath)
+		if err := ioutil.WriteFile(generatedFilePath, []byte(fileContent), 0600); err != nil {
+			return err
+		}
+		// Make sure the file has the permissions expected by the Postgresql driver, otherwise it will bail
+		if err := os.Chmod(generatedFilePath, 0600); err != nil {
+			return err
+		}
+		ds.JsonData.Set(jsonFieldName, generatedFilePath)
+		return nil
+	}
+
+	if generatedFilePath != "" {
+		s.logger.Debug("Deleting cert file since no content is provided", "path", generatedFilePath)
+		exists, err := fs.Exists(generatedFilePath)
+		if err != nil {
+			return err
+		}
+		if exists {
+			if err := os.Remove(generatedFilePath); err != nil {
+				return fmt.Errorf("failed to remove %q: %w", generatedFilePath, err)
+			}
+		}
+	}
+	ds.JsonData.Set(jsonFieldName, "")
+	return nil
+}
+
+func (s *postgresService) writeCertFiles(ds *models.DataSource) error {
+	s.logger.Debug("Writing TLS certificate files to disk")
+
+	decrypted := ds.SecureJsonData.Decrypt()
+	tlsRootCert := decrypted["tlsCACert"]
+	tlsClientCert := decrypted["tlsClientCert"]
+	tlsClientKey := decrypted["tlsClientKey"]
+
+	if tlsRootCert == "" && tlsClientCert == "" && tlsClientKey == "" {
+		s.logger.Debug("No TLS/SSL certificates provided")
+	}
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
```

Comment:
I think this might be problematic, basically holding a lock and making every postgres data source instance that uses `file-content` method wait here. Something we need to think a bit more about before being able to merge this.